modules/csg/csg.cpp

```diff
@@ -278,188 +273,195 @@ void CSGBrush::copy_from(const CSGBrush &p_brush, const Transform3D &p_xform) {
-		if (p_brush_b.faces[i].material != -1) {
-			material = p_brush_b.materials[p_brush_b.faces[i].material];
+	// runIndex needs an explicit end value.
+	mesh.runIndex.push_back(mesh.triVerts.size());
+
+	ERR_FAIL_COND_MSG(mesh.vertProperties.size() % mesh.numProp != 0, "Invalid vertex properties size.");
+
+	mesh.precision = p_snap;
+
+	/**
+	 * MeshGL64::merge(): updates the mergeFromVert and mergeToVert vectors in order to create a
+	 * manifold solid. If the MeshGL64 is already manifold, no change will occur, and
+	 * the function will return false.
+	 */
+	if (mesh.Merge()) {
+		std::vector<int32_t> index_map(mesh.vertProperties.size() / MANIFOLD_MAX, -1);
+		const size_t vertices_count = mesh.mergeFromVert.size();
+		for (size_t i = 0; i < vertices_count; ++i) {
+			index_map[mesh.mergeFromVert[i]] = mesh.mergeToVert[i];
 		}
-
-		if (build2DFaceCollection.build2DFacesB.has(i)) {
-			build2DFaceCollection.build2DFacesB[i].addFacesToMesh(mesh_merge, p_brush_b.faces[i].smooth, p_brush_b.faces[i].invert, material, true);
-		} else {
-			Vector3 points[3];
-			Vector2 uvs[3];
-			for (int j = 0; j < 3; j++) {
-				points[j] = p_brush_b.faces[i].vertices[j];
-				uvs[j] = p_brush_b.faces[i].uvs[j];
+		const size_t indices_count = mesh.triVerts.size();
+		for (size_t i = 0; i < indices_count; ++i) {
+			if (index_map[i] > -1) {
+				mesh.triVerts[i] = index_map[i];
 			}
-			mesh_merge.add_face(points, uvs, p_brush_b.faces[i].smooth, p_brush_b.faces[i].invert, material, true);
 		}
 	}
 
-	// Mark faces that ended up inside the intersection.
-	mesh_merge.mark_inside_faces();
-
-	// Create new brush and fill with new faces.
-	r_merged_brush.faces.clear();
-
-	switch (p_operation) {
-		case OPERATION_UNION: {
-			int outside_count = 0;
-
-			for (int i = 0; i < mesh_merge.faces.size(); i++) {
-				if (mesh_merge.faces[i].inside) {
-					continue;
-				}
-				outside_count++;
-			}
-
-			r_merged_brush.faces.resize(outside_count);
-
-			outside_count = 0;
-
-			for (int i = 0; i < mesh_merge.faces.size(); i++) {
-				if (mesh_merge.faces[i].inside) {
-					continue;
-				}
-
-				for (int j = 0; j < 3; j++) {
-					r_merged_brush.faces.write[outside_count].vertices[j] = mesh_merge.points[mesh_merge.faces[i].points[j]];
-					r_merged_brush.faces.write[outside_count].uvs[j] = mesh_merge.faces[i].uvs[j];
-				}
-
-				r_merged_brush.faces.write[outside_count].smooth = mesh_merge.faces[i].smooth;
-				r_merged_brush.faces.write[outside_count].invert = mesh_merge.faces[i].invert;
-				r_merged_brush.faces.write[outside_count].material = mesh_merge.faces[i].material_idx;
-				outside_count++;
-			}
-
-			r_merged_brush._regen_face_aabbs();
-
-		} break;
-
-		case OPERATION_INTERSECTION: {
-			int inside_count = 0;
+	r_manifold = manifold::Manifold(mesh);
+	manifold::Manifold::Error err = r_manifold.Status();
+	if (err != manifold::Manifold::Error::NoError) {
+		print_error(String("Manifold creation from mesh failed:" + itos((int)err)));
+	}
```

Comment:
Shouldn't this return a boolean or Error if it fails? Or is it impossible to trigger an error here (this is more like an assertion)

Might be good to use one of the ERR_ macros.

Reply:
If the geometry is not manifold able I prefer it disappears rather to crash godot. An error message every frame is not fun either.